### PR TITLE
Add first draft for blog post on fork/upstream relationship

### PR DESCRIPTION
**Pull request by Bisaloo**
Moving the conversation here for transparency. I briefly pitched this to Stef & Steffi already.

In terms of timeline:

- Hugo to finish draft (hopefully during the first week of December)
- Invite Stef to review & add context
- Share with all the community call participants to see if they have some extra thoughts to add

With end of year holidays coming up, this likely means this won't be finalized before early 2025.

Checklist:

* [x] I have read the Content Guidelines.
* [x] I have read the Technical Guidelines.
* [ ] I used or followed the R Markdown or Markdown template.
* [x] I have followed the Style Guide.
* [x] I created or updated my author metadata with correct folder name.
* [x] I have added relevant tags after browsing existing tags (including "community" tag).
* ~~I have added the "tech notes" tag if this is a technote.~~
* [x] I ran `roblog::ro_lint_md()` on index.md (optional).
* [ ] I ran `roblog::ro_check_urls()` on index.md (optional).
* [x] I ran a spell-check on index.md.

### CODE REVIEW

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
+---
+slug: "forks-upstream-relationship"
+title: The dynamic relationship of forks with their upstream repository
+date: '2024-12-08'
```
Comment (Bisaloo) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:4]: Determine authoring policy: everyone in the comm call or only whoever gives feedback on this draft?

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+---
+
+## Context setting
+
+rOpenSci organizes [monthly co-working hours](https://ropensci.org/coworking/) on a variety of topics.
+But the constant is the quality of the discussion that ensues and the renewed energy that comes from it.
+
+The November coworking session welcomed Stefanie Butland from [openscapes](https://www.openscapes.org/) as a co-host.
+She shared openscape's approach of ["Forking as a worldview"](https://rladiesrome.org/talks/2024/meetup/11182024/index.html)
+and invited participants to share their experience using forks to build on existing projects.
```
Comment (Bisaloo) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:14]: Stef, do you want to add more here?

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+## Facilitating the cross-pollination between forks and upstream
+
+In this section, we focus on the second type of forks, 
+where the fork author wants to stay in sync with the upstream repository.
+We mentioned that conflicts were a major downside of this approach.
+
+What are the strategies to minimize the conflicts? 
+And why is it in the interest of everyone (the fork owners, but also upstream maintainers) to minimize them?
+
+### Simplifying syncs and reducing git conflicts {#simplifying-syncs}
```
Comment (Bisaloo) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:55]: I think this section could benefit from code examples or illustrations. There are links to PRs but something in the post itself may also help.

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+---
+
+## Context setting
+
+rOpenSci organizes [monthly co-working hours](/coworking/) on a variety of topics.
+But the constant is the quality of the discussion that ensues and the renewed energy that comes from it.
+
+The November coworking session welcomed Stefanie Butland from [openscapes](https://www.openscapes.org/) as a co-host.
+She shared openscape's approach of ["Forking as a worldview"](https://rladiesrome.org/talks/2024/meetup/11182024/index.html)
+and invited participants to share their experience using forks to build on existing projects.
```
Comment (stefaniebutland) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:22]:
```suggestion
Stefanie invited participants to share their experiences in forking to adapt or build on existing projects, and we came back to some practical challenges of forking.
```

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+This blog post summarizes some of the insights that emerged during the subsequent discussion.
+
+## The different types of forks
+
+We identified two different types of forks, which take place in different contexts,
+and will have different relationships with their upstream repository in the long-run:
+
+1. Forks can be used as a starting point for a new diverging project, rather than starting from scratch.
+In this situation, the fork author does not intent to sync with the upstream repository after the initial fork.
```
Comment (stefaniebutland) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:32]:
```suggestion
In this situation, the fork author does not intend to sync with the upstream repository after the initial fork.
```

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+This type of fork may therefore miss out on new features developed in upstream.
+This type of fork will often happen when you copy template for a personal website (example shared by Will Gearty),
```
Comment (stefaniebutland) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:34]:
```suggestion
This type of fork will often happen when you copy a template for a personal website (example shared by Will Gearty), 
```

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+We identified two different types of forks, which take place in different contexts,
+and will have different relationships with their upstream repository in the long-run:
+
+1. Forks can be used as a starting point for a new diverging project, rather than starting from scratch.
+In this situation, the fork author does not intent to sync with the upstream repository after the initial fork.
+This type of fork may therefore miss out on new features developed in upstream.
+This type of fork will often happen when you copy template for a personal website (example shared by Will Gearty), 
+or when you re-use an existing project of yours as a boilerplate to build your new project.
+
+2. Forks can also be used to custom or tweak the original project,
```
Comment (stefaniebutland) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:37]:
```suggestion
2. Forks can also be used to customize or tweak the original project, 
```

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+This type of fork may therefore miss out on new features developed in upstream.
+This type of fork will often happen when you copy template for a personal website (example shared by Will Gearty), 
+or when you re-use an existing project of yours as a boilerplate to build your new project.
+
+2. Forks can also be used to custom or tweak the original project, 
+in a way that upstream cannot or doesn't want to address.
+In this type of forks, we usually try to not diverge too much from upstream to benefit from new features,
+and we regularly sync with them.
+A major downside is that syncing will often come with gnarly git conflicts.
+An example of this is various organizations forking the Carpentries workbench to customize the look and feel or their internal Carpentries-style training materials.
```
[the remark this answers is not in the document]
Comment (stefaniebutland) [reply content/blog/2024-12-01-forks-upstream-relationship/index.md:42]: First sentence of 2. should have similar structure to 1. where the main contrast is we don't want to diverge, and then say upstream source does not want to incorporate our adaptations.

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
```
Comment (stefaniebutland) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:42]:
```suggestion
An example of this is various organizations forking the Carpentries workbench to customize the look and feel of their internal Carpentries-style training materials.
```

Comment (stefaniebutland) [reply content/blog/2024-12-01-forks-upstream-relationship/index.md:42]: would be great to give a sense of how many orgs have done this. I thought it was pretty impressive. Shows the value of the open sharing of the original project

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+This is also what typically happens when open-source projects are packaged in Operating System distributions, such as Debian.
```
Comment (stefaniebutland) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:43]: Is this important to say? My non-software-dev mind doesn't get it. Does it fit the big picture we're trying to convey?

File: content/blog/2024-12-01-forks-upstream-relationship/index.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (stefaniebutland) [reply content/blog/2024-12-01-forks-upstream-relationship/index.md:42]: I also think it would be fantastic to link to your own forked version of this so folks can see what it looks like in action. ... haha I see that below.

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+But the upstream repository regularly receives updates with bug fixes or new features.
+
+The few first months were painful because of the conflicts that arose when syncing with upstream.
+But a couple of changes in upstream have greatly reduced the probability of conflicts.
+
+In a nutshell, upstream can make it easier for forks to stay in sync by reducing the number of files that a fork needs to modify,
+and by isolating these changes in specific locations.
+
+In practice, this can for example be done by using configuration files and variables.
+This serves the double purpose of being more DRY and of separating changes.
```
Comment (stefaniebutland) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:68]: what is DRY?

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+In this section, we focus on the second type of forks, 
+where the fork author wants to stay in sync with the upstream repository.
+We mentioned that conflicts were a major downside of this approach.
+
+What are the strategies to minimize the conflicts? 
+And why is it in the interest of everyone (the fork owners, but also upstream maintainers) to minimize them?
+
+### Simplifying syncs and reducing git conflicts {#simplifying-syncs}
+
+Going back to the example of organizations forking the Carpentries workbench,
```
Comment (stefaniebutland) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:57]: I'd love a sentence on why you did this! Comes back to the big picture of value of forking and value of Carpentries making the workbench open. A sentence on epiverse and what the unique training is? imo helps people picture when they might want to use the advice in this post

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+removing the need to change it in multiple locations.
+
+But from the fork owner perspective, some techniques can also make conflict resolution smoother. 
+A major location of conflicts are auto-generated changes or files.
+In the workbench example, this is the concatenated and minified CSS files.
+They tend to concentrate conflicts because they regroup changes from multiple files, 
+and they are minified on a single line, which means git cannot distinguish what was changed.
+
+But you can automate the conflict resolution for these cases.
+The key is to make automated changes in a separate commit.
```
Comment (stefaniebutland) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:79]: can you link to an example commit? So ppl can see and copy approach, rather than having to imagine how to do it

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+When this commit conflicts, you can re-run the script that generates the file,
+and commit the result, without having to think about what should be kept or not.
+
+Alternatively, upstream can sometimes stop tracking auto-generated files from git,
+and instead render them on the fly at build time.
+
+### Contributing back to upstream
+
+The previous sections highlighted how much work maintaining a fork.
```
Comment (stefaniebutland) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:88]:
```suggestion
The previous sections highlighted how much work goes into maintaining a fork.
```

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+Is it possible to integrate the features to upstream and avoid having to maintain a fork?
+
+Forks can be a good way to pilot new features or approaches.
+Note that this is not always possible as upstream may have a restricted scope, or limited resources for maintenance.
+
+Forks are sometimes intended from the start as short-lived.
+They only exist until the PR is merged when the user cannot wait for the PR to be merged, especially if upstream is slow to review PRs.
```
Comment (stefaniebutland) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:95]:
```suggestion
They only exist until the PR to upstream is merged when the user cannot wait for the PR to be merged, especially if upstream is slow to review PRs.
```

Comment (stefaniebutland) [reply content/blog/2024-12-01-forks-upstream-relationship/index.md:95]: needs some clarification

File: content/blog/2024-12-01-forks-upstream-relationship/index.md (hunk 14 above)
Comment (stefaniebutland) [reply content/blog/2024-12-01-forks-upstream-relationship/index.md:88]: This section could use some clarification  or condensing. I wasn't sure of the take-home message in a quick read

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+
+But in other cases, can forks that started as a way to customize a project and start accumulating new features contribute back to upstream?
+Forks can indeed be a great way to pilot new features, and to show the value of these features to upstream.
+
+When contributing back, each pull request should focus on a single feature or bug fix, rather than pushing all changes the fork has accumulated.
+It can also build confidence to show that you've been using this feature for a while without issues.
+
+### It takes two to tango
+
+Getting forks to contribute back is only possible is they can stay in touch with upstream.
```
Comment (stefaniebutland) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:106]:
```suggestion
Getting forks to contribute back is only possible if they can stay in touch with upstream.
```

Comment (stefaniebutland) [reply content/blog/2024-12-01-forks-upstream-relationship/index.md:106]: maybe add a comment on value of communicating your intent up front with the upsteam source? "I'd like to contribute back to upstream. Does the project have the capacity for this (merging PRs in a timely way). How could that work for you?"

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+Indeed, forks that have diverged too much from upstream will have a hard time contributing back, and making nice palatable pull requests.
+In other words, upstream repositories should follow the steps outlined in the ["Simplifying syncs and reducing git conflicts"](#simplifying-syncs) section to ensure forks can more easily stay in sync.
+
+This highlights the symbiotic relationship that can exist between forks and upstream repositories, when both parties are willing to make the effort.
+
+## Conclusion
+
+While sometimes presented as a fracture, forks are an integral part of the open-source ecosystem.
+They offer a way to build on existing projects, to add extra features without increasing the maintenance burden of the upstream repository, to pilot new features.
+They can be hard to maintain, but it is also possible for the fork and upstream maintainers to work together to make the process smoother.
```
Comment (stefaniebutland) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:116]: Maybe at bottom have a call to action asking people to consider what projects they are planning that could be forked. Code, lessons, documentation. Do they envision diverging or contributing back? With this info in the post how might they approach it.

**Review by stefaniebutland — COMMENTED** [review]
Hugo, so fantastic that you've written this up, and as usual, your ideas come across really well. I think it would be good if you're comfortable with it to centre yourself earlier in the post. Saying more about your motivation to fork the workbench (more in one of my comments). This can make it relatable to a broader audience where there's a real protagonist living through this, and coming up with a set of ideas. 
I'd love to do a final review before publishing. 
Thank you so much for the discussion and for doing this. And to @steffilazerte for hosting the awesome coworking sessions!!

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+
+In a nutshell, upstream can make it easier for forks to stay in sync by reducing the number of files that a fork needs to modify,
+and by isolating these changes in specific locations.
+
+In practice, this can for example be done by using configuration files and variables.
+This serves the double purpose of being more DRY and of separating changes.
+In the example of the Carpentries workbench, [a pull request added a Sass variable to change the font](https://github.com/carpentries/varnish/pull/151), 
+removing the need to change it in multiple locations.
+
+But from the fork owner perspective, some techniques can also make conflict resolution smoother.
```
Comment (steffilazerte) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:72]: What kind of technique? Was adding a Sass variable not a technique?

```suggestion
From the fork owner perspective, some other techniques can also make conflict resolution smoother.
```

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+When this commit conflicts, you can re-run the script that generates the file,
+and commit the result, without having to think about what should be kept or not.
+
+Alternatively, upstream can sometimes stop tracking auto-generated files from git,
+and instead render them on the fly at build time.
+
+### Contributing back to upstream
+
+The previous sections highlighted how much work maintaining a fork.
+Is it possible to integrate the features to upstream and avoid having to maintain a fork?
```
Comment (steffilazerte) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:89]:
```suggestion
So, is it perhaps possible to integrate the features to upstream and avoid having to maintain a fork?
```

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+
+Forks can be a good way to pilot new features or approaches.
```
Comment (steffilazerte) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:91]: Which can be then integrated back into the upstream?

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+Note that this is not always possible as upstream may have a restricted scope, or limited resources for maintenance.
+
+Forks are sometimes intended from the start as short-lived.
+They only exist until the PR is merged when the user cannot wait for the PR to be merged, especially if upstream is slow to review PRs.
+In this case, the fork is already contributing back to upstream.
```
Comment (steffilazerte) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:96]: How, if it's not merged?

File: content/blog/2024-12-01-forks-upstream-relationship/index.md (hunk 14 above)
Comment (steffilazerte) [reply content/blog/2024-12-01-forks-upstream-relationship/index.md:88]: Agreed, I could see some interesting nuggets of information, but wasn't clear on the main idea.

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+Forks can indeed be a great way to pilot new features, and to show the value of these features to upstream.
+
+When contributing back, each pull request should focus on a single feature or bug fix, rather than pushing all changes the fork has accumulated.
+It can also build confidence to show that you've been using this feature for a while without issues.
+
+### It takes two to tango
+
+Getting forks to contribute back is only possible is they can stay in touch with upstream.
+Indeed, forks that have diverged too much from upstream will have a hard time contributing back, and making nice palatable pull requests.
+In other words, upstream repositories should follow the steps outlined in the ["Simplifying syncs and reducing git conflicts"](#simplifying-syncs) section to ensure forks can more easily stay in sync.
```
Comment (steffilazerte) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:108]: Perhaps soften this, as it sounds like the onus is just on the Upstream maintainers, but that's not necessarily always the case. Sometimes they need to make big changes and sometimes it's more important to improve the project than to worry about too much divergence with the forks (or waiting for forks to submit their PRs before moving on). 

I like Stef's suggestion of communication. 

What do you think of something along these lines:
```suggestion
In other words, communication, and where possible, following the steps outlined in the ["Simplifying syncs and reducing git conflicts"](#simplifying-syncs) section are key to ensuring forks can more easily stay in sync.
```

**Review by steffilazerte — COMMENTED** [review]
Hi @Bisaloo,

Sorry it took me a bit longer to get to than I had hoped.
Thanks for writing this! Also for the lovely comments about coworking, that makes me very happy to hear!

I agree with Stef's comments in general and have added some of my own. My main suggestions are to clarify a bit the section "Contributing back to upstream" (as Stef suggested), and also to consider in what ways can the maintainers of *forks* work with upstream maintainers to keep things smooth? As you say, it takes two to tango, but most of your comments reflect what upstream can do differently. In particular, I really liked Stef's suggestion to communicate with the upstream maintainers. Alternatively, it's totally fine to focus on what upstream maintainers can do, but then I would perhaps reframe the context a bit at the start to make this clear.

* [X] post follows Content Guidelines
* [X] post follows Style Guide
* [X] title is in Title Case
* [ ] publication date is ok -> **We'll pick a date close to when you're ready, no hurry!**
* [ ] author metadata is provided with correct folder name -> **I'll check again when you have authors**
* [X] html not included in pull request of Rmd post
* [X] I ran `roblog::ro_lint_md()` on index.md
* [X] I ran `roblog::ro_check_urls()` on index.md
* [X] I ran a spell-check on index.md
* [X] YAML subject tags are ok ("tech notes" for tech notes; "community" for non-staff non-editor)
* [ ] YAML field 'editor' is filled with your name
* [X] YAML field 'preface' is present if necessary

**Comment by stefaniebutland** [comment]
Here just to say how happy it makes me to be in a PR review with @steffilazerte ❤️

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+
+## The different types of forks
+
+We identified two different types of forks, which take place in different contexts,
+and will have different relationships with their upstream repository in the long-run:
+
+1. Forks can be used as a starting point for a new diverging project, rather than starting from scratch.
+In this situation, the fork author does not intend to sync with the upstream repository after the initial fork.
+This type of fork may therefore miss out on new features developed in upstream.
+This type of fork will often happen when you copy a template for a personal website (an example shared by Will Gearty in our discussions),
```
Comment (Bisaloo) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:34]: TODO: add link to Will's online presence if he doesn't end up as an author

**Comment by stefaniebutland** [comment]
Confirming that we'd love to cross-post this on the Openscapes blog! I can copy directly from the GitHub source of the post.

**Comment by steffilazerte** [comment]
Awesome! As Hugo is on holidays and I'm taking holidays shortly, we decided not to rush this post and will pick it back up in the New Year.

File: content/blog/2024-12-01-forks-upstream-relationship/index.md (hunk 10 above)
Comment (Bisaloo) [reply content/blog/2024-12-01-forks-upstream-relationship/index.md:48]: I'm on the fence. 

I think it's a really good example because that's probably one of the most common cases when you have this kind of long-lived forks that are just here to patch some existing projects and make them fit into the specific operating system policies.
If anyone was curious about the kind of (social & technical) issues that can arise from these scenarios, they would definitely get a lot of value looking at this example.

At the same time, you are right it may only speak to a minority of users at this time.

The ideal outcome in my opinion would be to find a way to keep it but rephrase it so that it's clearer even for people unfamiliar with this ecosystem. Any suggestions?

Comment (stefaniebutland) [reply content/blog/2024-12-01-forks-upstream-relationship/index.md:50]: This extra context is really helpful! Def keep. 
> This is also what typically happens when open-source projects are packaged in Operating System distributions, such as Debian.

People who work in Linux operating systems will recognize .... (something something that maybe @steffilazerte could write - your comment above sounds super relatable)

**Comment by steffilazerte** [comment]
@yabellini, @mpadge, @estedeahora, @DaniMori, @RichardLitt, @rmk118, @willgearty!

@Bisaloo (with @stefaniebutland) wrote up this blog post for rOpenSci after being inspired by the conversation you all had in the Coworking Session "[Getting to know Openscapes](https://ropensci.org/events/coworking-2024-11/)" which you all attended in November. The conversation served as a catalyst to ideas which resulted in the post.

As such, Hugo wanted to invite you all to review the post and offer any feedback. In particular, is everything clear to you? Are there major misconceptions or important related topics that are ignored but should be discussed? We're not looking to do a complete reworking of the post, but want to make sure that we haven't neglected any important ideas or thoughts.

If you're happy with the post and agree, we'd also like to add you as authors (so please let us know if you agree).

If we could get feedback by **Wednesday Feb 12th 2025**, that would be great. If you'd really like to give feedback but that time frame doesn't work, let us know, otherwise we'll assume you have declined.

To review: 
- [See the draft post here](https://deploy-preview-886--ropensci.netlify.app/blog/2024/12/08/forks-upstream-relationship/)
- Add comments or suggestions (you can do a review) to this PR
- If you agree to be an author, let me know and I'll create an [author file](https://blogguide.ropensci.org/authortechnical.html#createauthorfile) for you

Also, there are two other participants Sergio Cabrera and John Ellis for whom we don't have contact info. If anyone has this, please let us know, and let us know if you think we have missed anyone.

Thank you!

[I will also be sending a copy of this via email as well to ensure everyone gets it.]

**Comment by RichardLitt** [comment]
This post doesn't cover the social side of forks, where they are often used as a means of collective bargaining (forking Node to ensure community owning the code that was improperly trademarked), power grabs (forking open source projects to lock down proprietary features in future developments), and other disagreements.

**Comment by stefaniebutland** [comment]
@RichardLitt those are excellent points. I'll defer to Hugo on this but I think this post is quite nicely focussed and the social examples you highlight would belong in a different post.

File: content/blog/2024-12-01-forks-upstream-relationship/index.md (hunk 10 above)
Comment (yabellini) [reply content/blog/2024-12-01-forks-upstream-relationship/index.md:50]: Add a link to something that people who want to know more about can follow.

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 24]
```diff
@@ -0,0 +1,150 @@
...
+
+The November coworking session welcomed Stefanie Butland from [openscapes](https://www.openscapes.org/) as a co-host.
+She shared Openscapes' approach of ["Forking as a worldview"](https://docs.google.com/presentation/d/1aL0TVFM7xxzTCJoE3tEArDE32wh59wkmBCy6Vcmwvy4/).
+Forks in the open-source software world refer to copies of an existing open-source project that add their own changes on top of the original codebase.
+Openscapes is extending this classic software development interpretation to an even more open philosophy of encouraging people to take anything that works to new places.
+Things like the value of sharing, reusing, and remixing of *how we work*, in addition to the work itself.
+Stefanie invited participants to share their experiences in forking to adapt or build on existing projects, and we came back to some practical challenges of forking.
+
+This blog post summarizes and develops some of the insights that emerged during the subsequent discussion.
+
```
Comment (yabellini) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:32]: I think it still needs to be clarified that the focus on the insight is in GitHub repos and code repositories.  Because of the previous explanation, I was expecting another kind of content that was not so code-related. It feels like a big jump from here to talk about upstream and repositories.

File: content/blog/2024-12-01-forks-upstream-relationship/index.md (hunk 17 above)
Comment (yabellini) [reply content/blog/2024-12-01-forks-upstream-relationship/index.md:150]: This is also related to sustainable environments; perhaps the core developers team doesn't want that new feature you need, but the fork allows you to have it anyway. I like the call for action.

**Comment by yabellini** [comment]
Love this post.  Thanks so much for writing it.  I made some very minor suggestions.  Feel free to ignore them.  You don't need to list me as an author; you did all the hard work. Thank you for asking anyway. 🙏

**Comment by mpadge** [comment]
Consider this also my official statement of encouragement for a uniquely important concept. I have no particular comments, and you also do not need to add me as an author. But thanks for considering regardless. Great work all!

File: content/blog/2024-12-01-forks-upstream-relationship/index.md (hunk 23 above)
Comment (willgearty) [reply content/blog/2024-12-01-forks-upstream-relationship/index.md:41]: https://williamgearty.com/

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 25]
```diff
@@ -0,0 +1,150 @@
...
+and will have different relationships with their upstream repository in the long-run:
+
+1. Forks can be used as a starting point for a new diverging project, rather than starting from scratch.
+In this situation, the fork author does not intend to sync with the upstream repository after the initial fork.
+This type of fork may therefore miss out on new features developed in upstream.
+This type of fork will often happen when you copy a template for a personal website (an example shared by Will Gearty in our discussions), 
+or when you re-use an existing project of yours as a boilerplate to build your new project.
+
+2. Forks can also be used to customize or tweak the original project, 
+in a way that upstream cannot or doesn't want to address.
```
Comment (willgearty) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:45]: It might be good to explain here that it isn't usually the case that upstream doesn't _want_ the customization, just that they don't have the capacity to implement them but are very much open to PRs (although it is sometimes the case that upstream has philosophical objections to particular customizations that they would not want a PR for).

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 26]
```diff
@@ -0,0 +1,150 @@
...
+
+The previous section highlighted how much work goes into maintaining a fork.
+And while we shared techniques to alleviate some of this work, regular syncing is still required, and conflicts are still going to come up from time to time.
+So, is it perhaps possible to integrate the features to upstream and avoid having to maintain a fork altogether?
+
+Forks are sometimes intended from the start as short-lived.
+They only exist until the PR to upstream is merged when the user cannot wait for the PR to be merged, especially if upstream is slow to review PRs.
+
+But in other cases, forks start as a way to customize a project and accumulate new features, without necessarily the intention to contribute them back initially.
+But it's possible for the fork maintainer to change them mind at any time and decide to contribute back.
```
Comment (willgearty) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:122]:
```suggestion
But it's possible for the fork maintainer to change their mind at any time and decide to contribute back.
```

**Comment by willgearty** [comment]
I enjoyed the read and added just a couple very minor comments! No need to include me as an author but I appreciate the shout-out within the text!

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 27]
```diff
@@ -0,0 +1,159 @@
...
+- coworking
+---
+
+## Context setting
+
+rOpenSci organizes [monthly co-working hours](/coworking/) on a variety of topics.
+But the constant is the quality of the discussion that ensues and the renewed energy that comes from it.
+
+The November coworking session welcomed Stefanie Butland from [openscapes](https://www.openscapes.org/) as a co-host.
+She shared Openscapes' approach of ["Forking as a worldview"](https://docs.google.com/presentation/d/1aL0TVFM7xxzTCJoE3tEArDE32wh59wkmBCy6Vcmwvy4/).
```
Comment (Bisaloo) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:26]: @stefaniebutland, it seems that this document is not readable by everyone.

**Comment by Bisaloo** [comment]
Thanks everyone for your valuable feedback!

---

> This post doesn't cover the social side of forks, where they are often used as a means of collective bargaining (forking Node to ensure community owning the code that was improperly trademarked), power grabs (forking open source projects to lock down proprietary features in future developments), and other disagreements.

I agree with Stef this probably deserves its own blog post. If you end up writing it, I would love to read it and link back to it.

@RichardLitt, would you like to be listed as author?

---

This PR is starting to be difficult to follow but I think I've addressed the various comments. 

@steffilazerte, do you confirm we should be good to go? We still need to adjust the date once you and @stefaniebutland agree on the date for cross-posting.

**Comment by RichardLitt** [comment]
I don't feel that I contributed very much. So, thank you, but I think I am alright. If you would like to add me, I guess, feel free?

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 28]
```diff
@@ -0,0 +1,161 @@
...
+- carpentries
+- coworking
+---
+
+## Context setting
+
+rOpenSci organizes [monthly co-working hours](/coworking/) on a variety of topics.
+But the constant is the quality of the discussion that ensues and the renewed energy that comes from it.
+
+The November coworking session welcomed Stefanie Butland from [openscapes](https://www.openscapes.org/) as a co-host.
```
Comment (stefaniebutland) [inline content/blog/2024-12-01-forks-upstream-relationship/index.md:25]:
```suggestion
The November coworking session welcomed Stefanie Butland from [Openscapes](https://www.openscapes.org/) as a co-host.
```

File: content/blog/2024-12-01-forks-upstream-relationship/index.md (hunk 27 above)
Comment (stefaniebutland) [reply content/blog/2024-12-01-forks-upstream-relationship/index.md:26]: updated so anyone with link can view

File: content/blog/2024-12-01-forks-upstream-relationship/index.md [hunk 29]
```diff
@@ -0,0 +1,108 @@
...
+They tend to concentrate conflicts because they regroup changes from multiple files, 
+and they are minified on a single line, which means git cannot distinguish what was changed.
+
+But you can automate the conflict resolution for these cases.
+The key is to make automated changes in a separate commit.
+When this commit conflicts, you can re-run the script that generates the file,
+and commit the result, without having to think about what should be kept or not.
+
+Alternatively, upstream can sometimes stop tracking auto-generated files from git,
+and instead render them on the fly at build time.
```
[the remark this answers is not in the document]
Comment (Bisaloo) [reply content/blog/2024-12-01-forks-upstream-relationship/index.md:111]: Yes, but I'm not completely sure when I will get around to doing it so I don't think it should be a blocker for this post.

**Comment by stefaniebutland** [comment]
Thursday Feb 20 to publish at Openscapes is great for us @steffilazerte !

**Comment by steffilazerte** [comment]
Great!
- @stefaniebutland what is the link we should use when cross linking?
- @Bisaloo could you change the dates in the folder name and YAML to the 20th?
- @Bisaloo here is the draft social media post (mastodon and linkedin), let me know if you have any comments/suggestions

```
[blog] The Dynamic Relationship of Forks with their Upstream Repository 

Back in November, rOpenSci coworking hosted a get-to-know @openscapes@fosstodon.org with Stefanie Butland who shared their "Forking as a worldview" approach. 

This sparked a discussion of forking among the attendees and inspired @grusonh@mastodon.social to write this blog post with feedback from the attendees! 

https://ropensci.org/blog/2024/12/08/forks-upstream-relationship/

Cross-posted on the Openscapes blog: https://XXXXX

🙏 Thanks to everyone who participated in the discussion!
```

**Comment by stefaniebutland** [comment]
@steffilazerte our link: https://openscapes.org/blog/2025-02-20-forks-upstream-relationship
Can you pls point me to a png/jpg of rOpenSci logo (hex?) I can use for a feature image? I'll put it beside Openscapes hex.

And oh gosh I love it when someone else drafts the social posts 😁 Thank you!

**Comment by steffilazerte** [comment]
Hi Stef, we don't use the Hex logo anymore, but we have [these logos](https://github.com/ropensci/logos). Perhaps one of the abbreviated ones, like [this one](https://camo.githubusercontent.com/3a17adfb590d79803ce2b5a1538ab3e315d8950e91cce7b48ec923b839784da0/68747470733a2f2f7261776769742e636f6d2f726f70656e7363692f6c6f676f732f6d61737465722f69636f6e5f73686f72745f636f6c6f722e737667).

**Comment by stefaniebutland** [comment]
Our [PR](https://github.com/Openscapes/website-new/pull/179) | [preview](https://deploy-preview-179--dreamy-solstice-e71879.netlify.app/blog/2025-02-20-forks-upstream-relationship/) is ready to publish and promo Thursday.

Y'all are awesome! Thanks so much for this

**Comment by steffilazerte** [comment]
Great thanks! Merged and advertised, thanks for all your hard work!
